### DISCUSSION TITLE
Switching from Laravel Deployer to the main Deployer project

### DISCUSSION THREAD

**Discussion by Jamesking56** [deployphp/deployer#2502] [Help needed] [answered]
Hi,

I'm trying to convert my Laravel projects from [Laravel Deployer](https://github.com/lorisleiva/laravel-deployer) to this main Deployer project. I am using `deployer/deploy` `v7.0.0-beta.20` and I'm having an issue with it.

<details>
<summary>deploy.php config file</summary>

```php
<?php

namespace Deployer;

require 'recipe/common.php';
require 'recipe/laravel.php';
require 'contrib/npm.php';

set('application', 'Laravel');
set('repository', '<snip>');
set('shared_dirs', ['public/.well-known', 'storage']);
set('upload_options', [
    'options' => [
        '--exclude=.git',
        '--exclude=.idea',
        '--exclude=node_modules',
        '--exclude=/.env',
        '--exclude=/.github',
        '--exclude=/config/deploy.php',
        '--exclude=/tests',
        '--exclude=/vendor',
        '--exclude=/.editorconfig',
        '--exclude=/.gitattributes',
        '--exclude=/.gitignore',
        '--exclude=/.php_cs',
        '--exclude=/audit-resolve.json',
        '--exclude=/deploy.php',
        '--exclude=/grumphp.yml.dist',
        '--exclude=/phpstan-baseline.neon',
        '--exclude=/phpstan.neon.dist',
        '--exclude=/phpunit.xml',
        '--exclude=/server.php',
        '--exclude=/webpack.mix.js',
    ],
]);

host('staging')
    ->setLabels(['stage' => 'staging'])
    ->setHostname('<snip>')
    ->setDeployPath('<snip>')
    ->setRemoteUser('<snip>')
    ->setPort(<snip>)
    ->set('keep_releases', 2);

task('deploy', [
    'deploy:prepare',
    'deploy:lock',
    'deploy:release',
    'upload',
    'deploy:shared',
    'artisan:storage:link',
    'artisan:view:cache',
    'artisan:config:cache',
    'artisan:migrate',
    'deploy:symlink',
    'artisan:queue:restart',
    'deploy:unlock',
    'deploy:cleanup',
    'deploy:publish',
]);

task('npm:run:production', function () {
    cd('{{release_or_current_path}}');
    run('npm run production');
});
task('upload', function () {
    upload(__DIR__ . "/", '{{release_path}}');
});

after('deploy:failed', 'deploy:unlock');

```
</details>

Deployer doesn't seem to run any of the commands in the `deploy` task I've created. `deploy.php` is in the root directory and from the root directory I am running `vendor/bin/dep deploy staging`.

<details>
<summary>Command Output</summary>

```
[staging] info deploying issue/595-switch-from-laravel-deployer-to-deployer-php                                                                                                                                                                                         
task deploy:setup
task deploy:lock
task deploy:release
task deploy:update_code
task deploy:shared
task deploy:writable
[staging]  RuntimeException  in writable.php on line 11:
[staging]
[staging]   Can't detect http user name.
[staging]   Please setup `http_user` config parameter.
[staging]
task deploy:failed
task deploy:unlock
```
</details>

As you can see, this task list in the output doesn't match the tasks I've put in the `deploy.php` file. Any ideas what I've done wrong here?

**Comment by lorisleiva** [comment, marked as answer]
Hi there 👋 

Yes I can see a few errors in your deploy.php file. 

First of all, the error that you get is due to the fact that Deployer cannot automatically identify the user that is running your HTTP server (either via Apache or Nginx). It does this by looking into relevant processes and finding a user which is not root.

To solve this you must explicitly set the `http_user` options — which could be `root` in your case. On the upgrade guide to 7, it provides an example where `http_user` is set to `false`. By looking at the code I’m struggling to find out if this would fix your issue but maybe try that as well haha.

Now to answer your question about tasks not matching. That’s because the `deploy:prepare` task is a task that simply groups a bunch of other tasks together. It basically invoke all tasks that prepare your release. The same applies for the `deploy:publish` task that invoke all tasks that work towards changing the `current` symlink to the release you’re deploying.

Concretely, these are the tasks they invoke.

```php

task('deploy:prepare', [
    'deploy:info',
    'deploy:setup',
    'deploy:lock',
    'deploy:release',
    'deploy:update_code',
    'deploy:shared',
    'deploy:writable',
]);

task('deploy:publish', [
    'deploy:symlink',
    'deploy:unlock',
    'deploy:cleanup',
    'deploy:success',
]);
```

As a result, that means your deployment flow currently repeats a lot of tasks.

If you want to see the full detail of tasks and nested tasks invoked in your `deploy` task, you may run `dep tree deploy` to output its task-tree.

One last little detail I’ve noticed: you don’t need to import `recipe/common.php`. The Laravel recipe already imports it for you.

I hope this helps. 🍀

**Reply by Jamesking56** [reply]
Thanks for the speedy reply @lorisleiva :+1:

**Reply by Jamesking56** [reply]
I've managed to get my deployment working thanks @lorisleiva :+1:
